File: backend/engine/indicators.py
"""
Technical indicator library.

All functions:
- Accept and return pandas Series with the same index as the input
- Return NaN for the warm-up period (no data dropping)
- Use float64 throughout
- Match TradingView calculations:
  - EMA uses ewm(span=N, adjust=False)  — recursive formula, not initialised
  - RSI/ATR/ADX use Wilder's RMA        — ewm(alpha=1/N, adjust=False)
  - MACD is EMA(fast) - EMA(slow), signal = EMA(macd_line)
  - BB uses simple std (ddof=1)
  - Stochastic %K = (close - lowest_low) / (highest_high - lowest_low) * 100

The warm-up difference vs TradingView (first ~50 bars) is negligible on
multi-year datasets and does not affect golden dataset test outcomes because
the fixture data slice is trimmed to start well past the warm-up period.
"""

import numpy as np
import pandas as pd


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _rma(series: pd.Series, period: int) -> pd.Series:
    """Wilder's Running Moving Average (RMA).  alpha = 1/period."""
    return series.ewm(alpha=1.0 / period, min_periods=period, adjust=False).mean()


# ---------------------------------------------------------------------------
# Public indicator functions
# ---------------------------------------------------------------------------

def ema(close: pd.Series, period: int) -> pd.Series:
    """
    Exponential Moving Average.
    Matches TradingView ta.ema(): ewm with span=period, adjust=False.
    """
    result = close.ewm(span=period, min_periods=period, adjust=False).mean()
    result.name = f"EMA_{period}"
    return result


def sma(close: pd.Series, period: int) -> pd.Series:
    """Simple Moving Average."""
    result = close.rolling(window=period, min_periods=period).mean()
    result.name = f"SMA_{period}"
    return result


def rsi(close: pd.Series, period: int = 14) -> pd.Series:
    """
    Relative Strength Index using Wilder's RMA smoothing.
    Matches TradingView ta.rsi().
    """
    delta = close.diff()
    gain = delta.clip(lower=0.0)
    loss = (-delta).clip(lower=0.0)

    avg_gain = _rma(gain, period)
    avg_loss = _rma(loss, period)

    # Avoid division by zero on all-loss or all-gain streaks
    rs = avg_gain / avg_loss.replace(0.0, np.nan)
    result = 100.0 - (100.0 / (1.0 + rs))
    # Handle edge cases: all gains → RSI=100, all losses → RSI=0
    result = result.where(avg_loss != 0, 100.0)
    result = result.where(avg_gain != 0, 0.0)
    result.name = f"RSI_{period}"
    return result


def macd(
    close: pd.Series,
    fast: int = 12,
    slow: int = 26,
    signal_period: int = 9,
) -> tuple[pd.Series, pd.Series, pd.Series]:
    """
    MACD: (fast EMA − slow EMA), signal EMA, histogram.

    Returns (macd_line, signal_line, histogram) — all pd.Series.
    """
    fast_ema = close.ewm(span=fast, min_periods=fast, adjust=False).mean()
    slow_ema = close.ewm(span=slow, min_periods=slow, adjust=False).mean()
    macd_line = fast_ema - slow_ema
    signal_line = macd_line.ewm(span=signal_period, min_periods=signal_period, adjust=False).mean()
    histogram = macd_line - signal_line

    macd_line.name = f"MACD_{fast}_{slow}_{signal_period}_line"
    signal_line.name = f"MACD_{fast}_{slow}_{signal_period}_signal"
    histogram.name = f"MACD_{fast}_{slow}_{signal_period}_hist"
    return macd_line, signal_line, histogram


def bollinger_bands(
    close: pd.Series,
    period: int = 20,
    std_dev: float = 2.0,
) -> tuple[pd.Series, pd.Series, pd.Series]:
    """
    Bollinger Bands: upper, middle (SMA), lower.

    Returns (upper, middle, lower) — all pd.Series.
    Uses ddof=1 (sample std), matching TradingView.
    """
    middle = close.rolling(window=period, min_periods=period).mean()
    rolling_std = close.rolling(window=period, min_periods=period).std(ddof=1)
    upper = middle + std_dev * rolling_std
    lower = middle - std_dev * rolling_std

    upper.name = f"BB_{period}_{std_dev}_upper"
    middle.name = f"BB_{period}_{std_dev}_middle"
    lower.name = f"BB_{period}_{std_dev}_lower"
    return upper, middle, lower


def atr(
    high: pd.Series,
    low: pd.Series,
    close: pd.Series,
    period: int = 14,
) -> pd.Series:
    """
    Average True Range using Wilder's RMA smoothing.
    Matches TradingView ta.atr().

    True Range = max(high-low, |high-prev_close|, |low-prev_close|)
    """
    prev_close = close.shift(1)
    tr = pd.concat(
        [high - low, (high - prev_close).abs(), (low - prev_close).abs()],
        axis=1,
    ).max(axis=1)
    result = _rma(tr, period)
    result.name = f"ATR_{period}"
    return result


def adx(
    high: pd.Series,
    low: pd.Series,
    close: pd.Series,
    period: int = 14,
) -> pd.Series:
    """
    Average Directional Index.  Returns only the ADX line (0-100).
    Matches TradingView ta.adx().
    """
    # True Range
    prev_close = close.shift(1)
    tr = pd.concat(
        [high - low, (high - prev_close).abs(), (low - prev_close).abs()],
        axis=1,
    ).max(axis=1)

    # Directional movement
    up_move = high - high.shift(1)
    down_move = low.shift(1) - low

    pos_dm = pd.Series(
        np.where((up_move > down_move) & (up_move > 0), up_move, 0.0),
        index=high.index,
    )
    neg_dm = pd.Series(
        np.where((down_move > up_move) & (down_move > 0), down_move, 0.0),
        index=high.index,
    )

    smooth_tr = _rma(tr, period)
    smooth_pos = _rma(pos_dm, period)
    smooth_neg = _rma(neg_dm, period)

    # DI+ and DI-
    di_pos = 100.0 * smooth_pos / smooth_tr.replace(0, np.nan)
    di_neg = 100.0 * smooth_neg / smooth_tr.replace(0, np.nan)

    # DX and ADX
    dx = 100.0 * (di_pos - di_neg).abs() / (di_pos + di_neg).replace(0, np.nan)
    result = _rma(dx, period)
    result.name = f"ADX_{period}"
    return result


def stochastic(
    high: pd.Series,
    low: pd.Series,
    close: pd.Series,
    k_period: int = 14,
    k_smooth: int = 3,
    d_period: int = 3,
) -> tuple[pd.Series, pd.Series]:
    """
    Stochastic Oscillator.

    Returns (%K_smoothed, %D) — both pd.Series, values 0-100.
    Matches TradingView ta.stoch().
    """
    lowest_low = low.rolling(window=k_period, min_periods=k_period).min()
    highest_high = high.rolling(window=k_period, min_periods=k_period).max()

    range_ = (highest_high - lowest_low).replace(0, np.nan)
    raw_k = 100.0 * (close - lowest_low) / range_

    # Smooth %K
    k = raw_k.rolling(window=k_smooth, min_periods=k_smooth).mean()
    d = k.rolling(window=d_period, min_periods=d_period).mean()

    k.name = f"STOCH_{k_period}_{k_smooth}_{d_period}_k"
    d.name = f"STOCH_{k_period}_{k_smooth}_{d_period}_d"
    return k, d

File: frontend/src/lib/indicators.ts
/**
 * Client-side technical indicator library.
 *
 * Formulas match the backend (engine/indicators.py) and TradingView:
 *  - EMA  : ewm(span=N, adjust=False)  → alpha = 2/(N+1)
 *  - RMA  : ewm(alpha=1/N, adjust=False) — used for RSI / ATR / ADX
 *  - BB   : sample std (ddof=1)
 *  - STOCH: SMA-smoothed %K then %D
 *
 * All functions return (number | null)[] — null for the warm-up period.
 */

export type Series = (number | null)[];

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Wilder's RMA: alpha = 1/period */
function rma(values: number[], period: number): Series {
  const alpha = 1 / period;
  const out: Series = new Array(values.length).fill(null);
  let val = values[0];
  for (let i = 0; i < values.length; i++) {
    if (i > 0) val = alpha * values[i] + (1 - alpha) * val;
    if (i >= period - 1) out[i] = val;
  }
  return out;
}

// ---------------------------------------------------------------------------
// Public indicator functions
// ---------------------------------------------------------------------------

export function ema(closes: number[], period: number): Series {
  const alpha = 2 / (period + 1);
  const out: Series = new Array(closes.length).fill(null);
  let val = closes[0];
  for (let i = 0; i < closes.length; i++) {
    if (i > 0) val = alpha * closes[i] + (1 - alpha) * val;
    if (i >= period - 1) out[i] = val;
  }
  return out;
}

export function sma(closes: number[], period: number): Series {
  const out: Series = new Array(closes.length).fill(null);
  for (let i = period - 1; i < closes.length; i++) {
    let sum = 0;
    for (let j = i - period + 1; j <= i; j++) sum += closes[j];
    out[i] = sum / period;
  }
  return out;
}

export function rsi(closes: number[], period = 14): Series {
  if (closes.length < 2) return new Array(closes.length).fill(null);
  const gains: number[] = [0];
  const losses: number[] = [0];
  for (let i = 1; i < closes.length; i++) {
    const d = closes[i] - closes[i - 1];
    gains.push(Math.max(d, 0));
    losses.push(Math.max(-d, 0));
  }
  const ag = rma(gains, period);
  const al = rma(losses, period);
  return closes.map((_, i) => {
    if (ag[i] === null || al[i] === null) return null;
    const g = ag[i] as number, l = al[i] as number;
    if (l === 0) return 100;
    if (g === 0) return 0;
    return 100 - 100 / (1 + g / l);
  });
}

export interface MacdResult { line: Series; signal: Series; hist: Series }

export function macd(
  closes: number[],
  fast = 12,
  slow = 26,
  signalPeriod = 9,
): MacdResult {
  const fastEma = ema(closes, fast);
  const slowEma = ema(closes, slow);
  const line: Series = closes.map((_, i) => {
    if (fastEma[i] === null || slowEma[i] === null) return null;
    return (fastEma[i] as number) - (slowEma[i] as number);
  });

  // EMA of the MACD line (recursive, starting from first non-null value)
  const alpha = 2 / (signalPeriod + 1);
  const signal: Series = new Array(closes.length).fill(null);
  let sigVal: number | null = null;
  let count = 0;
  for (let i = 0; i < closes.length; i++) {
    if (line[i] === null) continue;
    const v = line[i] as number;
    sigVal = sigVal === null ? v : alpha * v + (1 - alpha) * sigVal;
    count++;
    if (count >= signalPeriod) signal[i] = sigVal;
  }

  const hist: Series = line.map((v, i) => {
    if (v === null || signal[i] === null) return null;
    return (v as number) - (signal[i] as number);
  });

  return { line, signal, hist };
}

export interface BBResult { upper: Series; middle: Series; lower: Series }

export function bollingerBands(closes: number[], period = 20, stdDev = 2.0): BBResult {
  const mid = sma(closes, period);
  const upper: Series = new Array(closes.length).fill(null);
  const lower: Series = new Array(closes.length).fill(null);
  for (let i = period - 1; i < closes.length; i++) {
    const slice = closes.slice(i - period + 1, i + 1);
    const mean = slice.reduce((a, b) => a + b, 0) / period;
    // ddof=1 (sample std), matching TradingView and the Python backend
    const variance = slice.reduce((a, b) => a + (b - mean) ** 2, 0) / (period - 1);
    const std = Math.sqrt(variance);
    upper[i] = (mid[i] as number) + stdDev * std;
    lower[i] = (mid[i] as number) - stdDev * std;
  }
  return { upper, middle: mid, lower };
}

export function atr(highs: number[], lows: number[], closes: number[], period = 14): Series {
  const tr: number[] = [highs[0] - lows[0]];
  for (let i = 1; i < closes.length; i++) {
    tr.push(Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1]),
    ));
  }
  return rma(tr, period);
}

export function adx(highs: number[], lows: number[], closes: number[], period = 14): Series {
  const n = closes.length;
  const tr: number[] = [highs[0] - lows[0]];
  const posDM: number[] = [0];
  const negDM: number[] = [0];
  for (let i = 1; i < n; i++) {
    tr.push(Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1]),
    ));
    const up = highs[i] - highs[i - 1];
    const dn = lows[i - 1] - lows[i];
    posDM.push(up > dn && up > 0 ? up : 0);
    negDM.push(dn > up && dn > 0 ? dn : 0);
  }
  const sTr = rma(tr, period);
  const sPos = rma(posDM, period);
  const sNeg = rma(negDM, period);
  const dx: number[] = new Array(n).fill(0);
  for (let i = 0; i < n; i++) {
    if (!sTr[i]) continue;
    const t = sTr[i] as number;
    const diP = 100 * (sPos[i] as number) / t;
    const diN = 100 * (sNeg[i] as number) / t;
    const sum = diP + diN;
    if (sum > 0) dx[i] = 100 * Math.abs(diP - diN) / sum;
  }
  return rma(dx, period);
}

export interface StochResult { k: Series; d: Series }

export function stochastic(
  highs: number[],
  lows: number[],
  closes: number[],
  kPeriod = 14,
  kSmooth = 3,
  dPeriod = 3,
): StochResult {
  const n = closes.length;
  const rawK: Series = new Array(n).fill(null);
  for (let i = kPeriod - 1; i < n; i++) {
    const hh = Math.max(...highs.slice(i - kPeriod + 1, i + 1));
    const ll = Math.min(...lows.slice(i - kPeriod + 1, i + 1));
    rawK[i] = hh === ll ? null : 100 * (closes[i] - ll) / (hh - ll);
  }
  // %K = SMA(rawK, kSmooth)
  const k: Series = new Array(n).fill(null);
  for (let i = kPeriod + kSmooth - 2; i < n; i++) {
    const slice = rawK.slice(i - kSmooth + 1, i + 1);
    if (slice.some(v => v === null)) continue;
    k[i] = (slice as number[]).reduce((a, b) => a + b, 0) / kSmooth;
  }
  // %D = SMA(k, dPeriod)
  const d: Series = new Array(n).fill(null);
  for (let i = kPeriod + kSmooth + dPeriod - 3; i < n; i++) {
    const slice = k.slice(i - dPeriod + 1, i + 1);
    if (slice.some(v => v === null)) continue;
    d[i] = (slice as number[]).reduce((a, b) => a + b, 0) / dPeriod;
  }
  return { k, d };
}

// ---------------------------------------------------------------------------
// Helper: convert a Series + timestamps array → lightweight-charts data array
// ---------------------------------------------------------------------------
export function toChartData(
  times: number[],
  series: Series,
): { time: number; value: number }[] {
  const result: { time: number; value: number }[] = [];
  for (let i = 0; i < times.length; i++) {
    if (series[i] !== null && !isNaN(series[i] as number)) {
      result.push({ time: times[i], value: series[i] as number });
    }
  }
  return result;
}
